pkg/runtime/hotreload/differ/differ.go

```diff
@@ -0,0 +1,146 @@
+/*
+Copyright 2023 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package differ
+
+import (
+	"reflect"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	compapi "github.com/dapr/dapr/pkg/apis/components/v1alpha1"
+	httpendapi "github.com/dapr/dapr/pkg/apis/httpEndpoint/v1alpha1"
+	"github.com/dapr/dapr/pkg/components/secretstores"
+	"github.com/dapr/dapr/pkg/runtime/meta"
+	"github.com/dapr/dapr/pkg/runtime/wfengine"
+)
+
+type Resource interface {
+	compapi.Component | httpendapi.HTTPEndpoint
+	meta.Resource
+}
+
+type Result[T Resource] struct {
+	Deleted []T
+	Updated []T
+	Created []T
+}
+
+type LocalRemoteResources[T Resource] struct {
+	Local  []T
+	Remote []T
+}
+
+// Diff returns the difference between the local and remote resources of the
+// given kind.
+func Diff[T Resource](resources *LocalRemoteResources[T]) *Result[T] {
+	if resources == nil ||
+		(len(resources.Local) == 0 && len(resources.Remote) == 0) {
+		return nil
+	}
+
+	// missing are the resources which exist remotely but which don't exist
+	// locally or have changed.
+	missing := detectDiff(resources.Local, resources.Remote, nil)
```

Comment:
Could we return missing and deleted in one loop in `detectDiff`?

Reply:
We have to loop twice because we need to find the intersection, as well as the difference between the two sets.